### PR TITLE
Visible Import Types retrieval

### PR DESCRIPTION
**Pull request by elatoskinas**
# Visible Type Extraction
Continuation of https://github.com/mir-am/ML4SE/pull/3
Resolves: https://github.com/mir-am/ML4SE/issues/2

Adds a file that allows to do the following:
* Retrieve a list of all import statements of a file (including transitive imports)
* Retrieve a list of all (resolved) modules for the given file that it imports
* Retrieve a list of visible types that originate from imports for the given file

## Changes from original PR
* Import types now get resolved to their qualified names. So if we have a type originating from scikit.linalg, the type name will then now be "scikit.linalg"
* Imports are now checked recursively. We check imports of imports
* Refactoring changes to make it easier to use
* Minor bug fixes (mainly addressing false negatives for types in some cases)

## Usage & Integration
The type extractor is not yet integrated into the pipeline. Integration is straight forward:
1. Suppose we have an input file (path) `file`
2. Create a new `ModuleExtractor`
3. Call `ModuleExtractor.getTypes(file)`
4. The returned result is a set of types. We can then keep it as a set, or convert it to a list. This is our final result output.

### CODE REVIEW

**Comment by elatoskinas** [comment]
@mir-am The implemented extractor is in `module_type_extractor.py`. All other files are there solely for testing purposes, and we will remove them once we finalize this PR.

File: import_types_temp/module_type_extractor.py [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+            # Retrieve the package given by the import name. We have to split by the '.'
+            # to conform to the function call's signature.
+            # Note that the package will be found, since the _resolve_import method above
+            # will load the module as well, and if a path is returned, we will get the package.
+            i, module = import_finder._find_package(import_name.split("."))
+            return module
+        else:
+            return None
+
+    def is_from_import_type(self, import_name, import_from):
```
Comment (elatoskinas) [inline import_types_temp/module_type_extractor.py]: TODO: The method signature & description is outdated. This should be updated to reflect that it actually now returns the type.

**Comment by mir-am** [comment]
@elatoskinas
Thanks for the work.
Do you think that we can run the import tool for 5,000 Python projects? Considering the issues we discussed so far. Moreover, I will train the model of TypeWriter with the visible types that you sent.

**Comment by elatoskinas** [comment]
@mir-am Yes, we can certainly train it on 5000 repos; I'll get to this soon!

File: import_type_extraction/module_type_generator.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        and saves them to a single CSV file.
+
+        :param: out_dir  Location to output CSV to (including filename)
+        :param: prefix_filters - List of prefixes to filter out for resulting members in final CSV.
+        """
+        # Create missing dirs if needed
+        os.makedirs(out_dir, exist_ok=True)
+
+        DATA_FILES = list_files(self.output_dir)
+        df = parse_df(DATA_FILES, batch_size=128)
```
Comment (elatoskinas) [inline import_type_extraction/module_type_generator.py:279]: TODO: This should be done in chunks to accommodate for dataframes not fitting in memory

**Comment by elatoskinas** [comment]
Closing, moved to repo: https://github.com/saltudelft/visible-type-extractor

**Pull request closed without merging**